scripts/internal/workflow-helpers/ghtransport.go

```diff
@@ -0,0 +1,157 @@
+package workflow_helpers
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/google/go-github/v45/github"
+)
+
+/*
+Code based on: https://github.com/hashicorp/terraform-provider-github/blob/fa73654b66e37b1fd8d886141d9c2974e24ba42f/github/transport.go#L42-L109
+Sourced via: https://github.com/google/go-github/issues/431
+
+MIT License
+
+Copyright (c) 2020 GitHub
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+const (
+	ctxId      = "id"
+	writeDelay = 720 * time.Millisecond // https://github.com/google/go-github/issues/431#issuecomment-248767702
+)
+
+// rateLimitTransport implements GitHub's best practices
+// for avoiding rate limits
+// https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits
+type rateLimitTransport struct {
+	transport        http.RoundTripper
+	delayNextRequest bool
+	responseBody     []byte
+	accessToken      string
+
+	m sync.Mutex
+}
+
+func (rlt *rateLimitTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", "Bearer "+rlt.accessToken)
+	// Make requests for a single user or client ID serially
+	// This is also necessary for safely saving
+	// and restoring bodies between retries below
+	rlt.lock(req)
+
+	// If you're making a large number of POST, PATCH, PUT, or DELETE requests
+	// for a single user or client ID, wait at least one second between each request.
+	if rlt.delayNextRequest {
+		logging.Debug("Sleeping %s between write operations", writeDelay)
+		time.Sleep(writeDelay)
+	}
+
+	rlt.delayNextRequest = isWriteMethod(req.Method)
+
+	resp, err := rlt.transport.RoundTrip(req)
+	if err != nil {
+		rlt.unlock(req)
+		return resp, err
+	}
```

Comment:
Maybe `defer rlt.unlock(req)` instead.

Reply:
It ensures that the unlock is called. I think there is a return in this function that does not call unlock. It also means that future work does not have to step so carefully.

Reply:
Thing is, this code is taken from https://github.com/hashicorp/terraform-provider-github/blob/fa73654b66e37b1fd8d886141d9c2974e24ba42f/github/transport.go

It was authored about 3 years ago and seems to have been "stable" since then (judging by no further commits alone). Not understanding the full implications of unlocking the req I'd rather use it as is, as there might've been a reason for it.

Add to that the fact that this is only for our automations and not for any code we ship, I feel like it's safer to leave it as is.